### PR TITLE
Update MP config to 1.3 

### LINKED ISSUE #432: Microprofile Config 1.3
Upgrade MP Config to 1.3 to comply with MP 2.2 specification.

### PR DESCRIPTION
**Pull request by batsatt**
Fixes #432 and includes changes to SE for improved symmetry.

Unfortunately, the different models for handling config sources in SE and MP have made symmetry for environment variables a bit complicated. 

In SE, the contents of the config sources are _merged_ during the config build and not consulted during a search. In order to support environment variables taking precedence over application config entries, an optimistic mapping must take place **prior** to the merge; see `EnvironmentVariables.expand()`. This approach produces key aliases that are highly _likely_ to match search keys, but is limited in what it can do.

I've been referring to this approach as the "static" mapping.

In MP, the config sources _are_ consulted during the search and the `MpcSourceEnvironmentVariables` class can easily implement the mapping required by the [spec](https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/configsources.asciidoc). Prior to the current change, it was (mostly) sufficient to just do the same static mapping; however, the one place where this falls down is with unusual characters in keys (e.g. slash in `app.my/property`, or a Unicode character). The spec wants _any_ non-alphanumeric character other than underscore mapped to underscore, and the 1.3 config TCK checks for this case. 

To fix this, the current change switches to use a "dynamic" mapping within `MpcSourceEnvironmentVariables`; see its use of `EnvironmentVariableAliases`.

So far, so good, except... SE does not handle unusual characters, so it isn't symmetric in this case. We can't simply switch to the static mapping or else mapped environment variables can't take precedence. We can't enhance the static mapping because the set of unusual characters is vast and would obviously result in an unacceptable explosion of entries.

So we can either live with the asymmetry or try to address it. I've attempted to address it here, using the dynamic mapping as a last resort when a search **fails**; see `aliasGenerator` usage in `ConfigFactory`. If the config key with the odd characters is actually present in the config, you get that value; if it is not, but one of the MP aliases is present you get that. Note a few things here:

1. The value with the aliased key is not restricted to the env variables source (e.g. you could have `"FOO_BAR"` in an application source and find it using `"foo.bar"` as the key.
1. Since this lookup happens only after searching everything else, it doesn't support overriding an existing value.
1. This behavior is enabled only when the default environment variable source is enabled.
1. It adds a small cost for every lookup miss, but it is one-time since the result is cached.

Like I said... complicated. It is clearly debatable whether this fallback search is worth it.

### CODE REVIEW

**Comment by tomas-langer** [comment]
I would vote for asymmetry here.
Helidon SE config is supposed to be a snapshot of the config tree.
MP config is a dynamic map.

**Comment by batsatt** [comment]
Just to be clear: this asymmetry would mean that it isn't possible to use environment variables to override config values in SE with such characters in the key; it is this use case that seems potentially important. OTOH, I don't love the implementation weirdness, so I can go either way. 

At the moment, I'm focusing on correctly supporting converters, and will get back to this.

**Comment by tjquinno** [comment]
I'm a little late to the comment stream here.

Earlier, Bryan, you described what you called the "static" and "dynamic" techniques for handling "weird" characters in keys, and that an option to maintain symmetric behavior between SE and MP would be to introduce dynamic weird character handling into SE.

Tomas, your comment about "SE config is supposed to be a snapshot of the config tree" to me at least sounds as if you have interpreted Bryan's description of the alternative to mean it would introduce dynamic _config behavior_ into SE. 

IIUC that _is not_ what is suggested here. Rather, it is just when and how the handling of special characters in keys is handled.

**Comment by tomas-langer** [comment]
I have gone through the changes and it looks good. 
There are a few comments (in review).
I agree with what you said Bryan about the `hasConverter` and your solution seems to handle that nicely (and as I hope I understood correctly from the code without backward incompatible changes).

Please correct me if I am wrong, the only backward incompatibility is in a case where a class has a public default constructor and no public setters (as in previous version, we would just create an instance, now we fail), right?

File: microprofile/config/config/pom.xml [hunk 1]
```diff
@@ -54,6 +54,18 @@
...
+        <dependency>
+            <groupId>io.helidon.config</groupId>
+            <artifactId>helidon-config-test-infrastructure</artifactId>
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>javax.inject</groupId>
+            <artifactId>javax.inject</artifactId>
+            <version>1</version>
```
[the remark this answers is not in the document]
Comment (batsatt) [reply microprofile/config/config/pom.xml]: Huh, good catch. This seems to have been added by IntelliJ at some point before I did some refactoring. It isn't needed at all and I'll remove.

File: microprofile/config/config/src/test/java/io/helidon/microprofile/config/MpConfigTest.java [hunk 2]
```diff
@@ -36,7 +41,7 @@
 /**
  * Unit test for {@link MpConfig}.
  */
-class MpConfigTest {
+public class MpConfigTest {
```
[the remark this answers is not in the document]
Comment (batsatt) [reply microprofile/config/config/src/test/java/io/helidon/microprofile/config/MpConfigTest.java]: Oh, this is a template issue; I've fixed both.

**Comment by batsatt** [comment]
Yes, you're correct about that backward incompatibility. (Supporting that case seems wrong anyway; what is the point of a "conversion" that ignores the input?)

File: microprofile/config/config/src/main/java/io/helidon/microprofile/config/MpConfig.java [hunk 3]
```diff
@@ -287,34 +268,45 @@ public boolean hasConverter(Class<?> clazz) {
      */
     @SuppressWarnings("unchecked")
     public <T> T convert(Class<T> type, String value) {
-        if (null == value) {
+        if (value == null) {
```
[the remark this answers is not in the document]
Comment (batsatt) [reply microprofile/config/config/src/main/java/io/helidon/microprofile/config/MpConfig.java]: I was re-writing the method, and, sorry, but this idiom offends my sense of symmetry; made the change without thinking much about it. (Though I see I wasn't consistent about this just below; this one jumped out at me being alone at the top.)

I see lots of code in the project that uses the 'normal' comparison order; clearly this is a matter of taste. But, sure, I'll revert it.

File: microprofile/config/config/src/main/java/io/helidon/microprofile/config/MpConfig.java [hunk 4]
```diff
@@ -287,34 +268,45 @@ public boolean hasConverter(Class<?> clazz) {
...
-                    .build();
-
-            try {
-                return c.get("key").as(type).get();
-            } catch (ConfigMappingException e) {
-                throw new IllegalArgumentException("Failed to convert " + value + " to " + type.getName(), e);
-            }
+        // Use a local converter for this class if we have one
+        final Converter<?> converter = converters.get(type);
+        if (null != converter) {
```
[the remark this answers is not in the document]
Comment (batsatt) [reply microprofile/config/config/src/main/java/io/helidon/microprofile/config/MpConfig.java]: I prefer to put expected pathways first, and don't see a simple negation like this as complex.

In this case, we're explicitly trying to use a local converter if it is present so having the code that uses it first seems far more logical to me.

File: microprofile/config/config/src/main/java/io/helidon/microprofile/config/MpConfig.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomas-langer) [reply microprofile/config/config/src/main/java/io/helidon/microprofile/config/MpConfig.java]: > Yes, you're correct about that backward incompatibility. (Supporting that case seems wrong anyway; what is the point of a "conversion" that ignores the input?)

Yes, I agree and in that case I do not think this is a real backward incompatible change, rather a bugfix ;)

**Comment by tomas-langer** [comment]
Yes, I agree and in that case I do not think this is a real backward incompatible change, rather a bugfix ;)

**Pull request merged**